### DISCUSSION TITLE
What is the best way to use RPC for type safety in more complex app structures?

### DISCUSSION THREAD

**Discussion by banjo** [honojs/hono#1382] [Q&A]
Hi,

All the examples for [RPC](https://hono.dev/guides/rpc) only shows one route as an example. What is the best way to implement this with multiple routes? Or even, multiple app instances used as routes?

Like this:

```ts
const app = new Hono();

app.route("/feed", feed);
app.route("/item", item);
app.route("/items", items);

// where each app has several routes
const feed = new Hono();
feed.get(//...)
feed.post(//...)
```

The documentation doesn't really mention anything about this. Do we need to handle the type for each route and export them manually or what is the best way?

**Comment by yusukebe** [comment]
Indeed, I believe that separating "apps" and then combining them using `app.route()` is the optimal approach. And, if you want to use "RPC", try the following structure:

```ts
const feed = new Hono();
const feedRoutes = feed.get('/foo', (c) => c.jsonT(0)).post('/bar', (c) => c.jsonT('a'));

const item = new Hono();
const itemRoutes = item.get('/foo', (c) => c.jsonT(0)).post('/bar', (c) => c.jsonT('a'));

const mainApp = new Hono();

const allRoutes = mainApp.route('/feed', feedRoutes).route('/item', itemRoutes);
```

For more insights, @subhendupsingh has written a fantastic guide on the topic. Check it out!: 
https://dev.to/subhendupsingh/typed-fetch-with-sveltekit-and-hono-using-rpc-2clf

**Reply by banjo** [reply]
Amazing, thank you very much.

**Comment by alpharder** [comment]
Just use tRPC instead.

**Reply by yusukebe** [reply]
You are right and we have a [tRPC adapter](https://github.com/honojs/middleware/tree/main/packages/trpc-server), but we don't talk about it now.
